src/test/crimson/test_messenger_thrash.cc

```diff
@@ -0,0 +1,683 @@
+using crimson::common::local_conf;
+
+struct Payload {
+  enum Who : uint8_t {
+    PING = 0,
+    PONG = 1,
+  };
+  uint8_t who = 0;
+  uint64_t seq = 0;
+  bufferlist data;
+
+  Payload(Who who, uint64_t seq, const bufferlist& data)
+    : who(who), seq(seq), data(data)
+  {}
+  Payload() = default;
+  DENC(Payload, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.who, p);
+    denc(v.seq, p);
+    denc(v.data, p);
+    DENC_FINISH(p);
+  }
+};
+WRITE_CLASS_DENC(Payload)
+
+std::ostream& operator<<(std::ostream& out, const Payload &pl)
+{
+  return out << "reply=" << pl.who << " i = " << pl.seq;
+}
+
+
+namespace {
+
+seastar::logger& logger() {
+  return crimson::get_logger(ceph_subsys_test);
+}
+
+std::random_device rd;
+std::default_random_engine rng{rd()};
+std::uniform_int_distribution<> prob(0,99);
+bool verbose = false;
+
+entity_addr_t get_server_addr() {
+  static int port = 16800;
+  ++port;
+  entity_addr_t saddr;
+  saddr.parse("127.0.0.1", nullptr);
+  saddr.set_port(port);
+  return saddr;
+}
+
+uint64_t get_nonce() {
+  static uint64_t nonce = 1;
+  ++nonce;
+  return nonce;
+}
+
+struct thrash_params {
+  std::size_t servers;
+  std::size_t clients;
+  std::size_t connections;
+  std::size_t random_op;
+};
+
+class SyntheticWorkload;
+
+class SyntheticDispatcher final
+    : public crimson::net::Dispatcher {
+  public:
+  bool is_server;
+  bool got_new;
+  bool got_remote_reset;
+  bool got_connect;
+  std::map<crimson::net::ConnectionRef, std::list<uint64_t> > conn_sent;
+  std::map<uint64_t, bufferlist> sent;
+  unsigned index;
+  SyntheticWorkload *workload;
+
+  SyntheticDispatcher(bool s, SyntheticWorkload *wl):
+      is_server(s), got_new(false),
+      got_remote_reset(false), got_connect(false), index(0), workload(wl) {
+  }
+
+  std::optional<seastar::future<>> ms_dispatch(crimson::net::ConnectionRef con,
+                                               MessageRef m) {
+    if (verbose) {
+      logger().warn("{}: con = {}", __func__, con);
+    }
+    // MSG_COMMAND is used to disorganize regular message flow
+    if (m->get_type() == MSG_COMMAND) {
+      return seastar::now();
+    }
+
+    Payload pl;
+    auto p = m->get_data().cbegin();
+    decode(pl, p);
+    if (pl.who == Payload::PING) {
+      logger().info(" {} conn= {} {}", __func__,
+        m->get_connection(), pl);
+      got_new = true;
+      return reply_message(m, pl);
+    } else {
+      if (sent.count(pl.seq)) {
+        logger().info(" {} conn= {} {}", __func__,
+          m->get_connection(), pl);
+        ceph_assert(conn_sent[m->get_connection()].front() == pl.seq);
+        ceph_assert(pl.data.contents_equal(sent[pl.seq]));
+        conn_sent[m->get_connection()].pop_front();
+        sent.erase(pl.seq);
+      }
+      got_new = true;
+      return seastar::now();
+    }
+  }
+
+  void ms_handle_accept(crimson::net::ConnectionRef conn) {
+    for (auto& it : conn_sent[conn]) {
+      sent.erase(it);
+    }
+    conn_sent.erase(conn);
```

Comment:
Following classic OSD matching unit test logic.
```
void ms_handle_fast_connect(Connection *con) override {
    std::lock_guard l{lock};
    list<uint64_t> c = conn_sent[con];
    for (list<uint64_t>::iterator it = c.begin();
         it != c.end(); ++it)
      sent.erase(*it);
    conn_sent.erase(con);
    got_connect = true;
    cond.notify_all();
  }
  void ms_handle_fast_accept(Connection *con) override {
    std::lock_guard l{lock};
    list<uint64_t> c = conn_sent[con];
    for (list<uint64_t>::iterator it = c.begin();
         it != c.end(); ++it)
      sent.erase(*it);
    conn_sent.erase(con);
    cond.notify_all();
  }
```

Reply:
Strictly speaking, we only reference the original logic, instead of copying it without knowing why.

IIUC with crimson msgr, ms_handle(remote)_reset should be reliable to clean up the stale status across connection sessions, there should be nothing to do with ms_handle_accept (except to check the status is clean) unless we are missing someting.

For ms_handle_connect, I believe it just notifies that connection is established, but the connection is able to queue messages as soon as it is created, so no need to clear the pending messages.

I think I'm good with this as it is already working, and we can guadually improve the test in the follow-up PRs.

Reply:
Both methods are updated now to **only** notify that the connection is established/accepted.